src/Network/Minio/PresignedOperations.hs

```diff
@@ -0,0 +1,281 @@
+
+-- | Set the content-type header for the upload.
+ppCondContentType :: Text -> PostPolicyCondition
+ppCondContentType = PPCEquals "Content-Type"
+
+-- | Set the object name constraint for the upload.
+ppCondKey :: Object -> PostPolicyCondition
+ppCondKey = PPCEquals "key"
+
+-- | Set the object name prefix constraint for the upload.
+ppCondKeyStartsWith :: Object -> PostPolicyCondition
+ppCondKeyStartsWith = PPCStartsWith "key"
+
+-- | Status code that the S3-server should send on a successful POST
+-- upload
+ppCondSuccessActionStatus :: Int -> PostPolicyCondition
+ppCondSuccessActionStatus n =
+  PPCEquals "success_action_status" (show n)
+
+-- | This function creates a PostPolicy after validating its
+-- arguments.
+newPostPolicy :: UTCTime -> [PostPolicyCondition]
+              -> Either PostPolicyError PostPolicy
+newPostPolicy expirationTime conds
+  -- object name condition must be present
+  | not $ any (keyEquals "key") conds =
+      Left PPEKeyNotSpecified
+
+  -- bucket name condition must be present
+  | not $ any (keyEquals "bucket") conds =
+      Left PPEBucketNotSpecified
+
+  -- a condition with an empty key is invalid
+  | any (keyEquals "") conds || any isEmptyRangeKey conds =
+      Left PPEConditionKeyEmpty
+
+  -- invalid range check
+  | any isInvalidRange conds =
+      Left PPERangeInvalid
+
+  -- all good!
+  | otherwise =
+      return $ PostPolicy expirationTime conds
+
+  where
+    keyEquals k' (PPCStartsWith k _) = k == k'
+    keyEquals k' (PPCEquals k _) = k == k'
+    keyEquals _ _ = False
+
+    isEmptyRangeKey (PPCRange k _ _) = k == ""
+    isEmptyRangeKey _ = False
+
+    isInvalidRange (PPCRange _ mi ma) = mi < 0 || mi > ma
+    isInvalidRange _ = False
+
+-- | Convert Post Policy to a string (e.g. for printing).
+showPostPolicy :: PostPolicy -> ByteString
+showPostPolicy = toS . Json.encode
+
+-- | Generate a presigned URL and POST policy to upload files via a
+-- browser. On success, this function returns a URL and a POST
+-- form-data.
+presignedPostPolicy :: PostPolicy
+                    -> Minio (ByteString, Map.Map Text ByteString)
+presignedPostPolicy p = do
+  ci <- asks mcConnInfo
+  signTime <- liftIO $ Time.getCurrentTime
+
+  let
+    extraConditions =
+      [ PPCEquals "x-amz-date" (toS $ awsTimeFormat signTime)
+      , PPCEquals "x-amz-algorithm" "AWS4-HMAC-SHA256"
+      , PPCEquals "x-amz-credential"
+        (T.intercalate "/" [connectAccessKey ci,
+                            decodeUtf8 $ mkScope signTime region])
+      ]
+    ppWithCreds = p {
+      conditions = conditions p ++ extraConditions
+      }
+    signData = signV4PostPolicy (showPostPolicy ppWithCreds)
+               signTime ci
+
+    -- compute form-data
+    mkPair (PPCStartsWith k v) = Just (k, v)
+    mkPair (PPCEquals k v) = Just (k, v)
+    mkPair _ = Nothing
+    formFromPolicy = Map.map toS $ Map.fromList $ catMaybes $ map mkPair $
+                     conditions ppWithCreds
```

Comment:
Use `<$>` instead of `Prelude.map` using Functor instance of `[a]`. Makes for good reading too :)
`formFromPolicy = Map.map toS $ Map.fromList $ catMaybes $ mkPair <$> conditions ppWithCreds`

Reply:
Done!